app/api.py

```diff
@@ -141,3 +157,54 @@ async def unsubscribe_to_track(track_namespace: str, subscriber: str):
                             detail="Relay is not in track namespace")
 
     track.remove_subscriber(subscriber)
+
+@app.get("/origin/{relayid}/{namespace}")
+async def get_origin(relayid: str, namespace: str):
+    nodes = network.nodes(data=True)
+    if len(nodes) < int(relayid):
+        raise HTTPException(status_code=response.status_code, detail="no such relay")
+
+    relay_in = None
+    for i, node in enumerate(nodes, start=1):
+        if (str(i) == relayid) & (relay_in == None):
+            relay_in = node[0]
+
+    response = await subscribe_to_track(track_namespace=namespace, subscriber=relay_in)
+    relay_out  = None
+    for i, node in enumerate(nodes, start=1):
+        if (node[0] == response) & (relay_out == None):
+            relay_out = i
+
+    if response != None:
+        response_json = {"url": f"https://10.3.0.{relay_out}/"}
+        return JSONResponse(content=response_json, headers={"Cache-Control": "no-cache, no-store, must-revalidate"})
+    else:
+        raise HTTPException(status_code=response.status_code, detail=response.text)
+
+@app.post("/origin/{relayid}/{namespace}", status_code=status.HTTP_200_OK)
+async def set_origin(relayid: str, namespace: str, origin: Annotated[Origin, Body()]):
+    nodes = network.nodes(data=True)
+    if len(nodes) < int(relayid):
+        raise HTTPException(status_code=response.status_code, detail="no such relay")
+
+    async with httpx.AsyncClient():
+        delay_budget = namespace.split("_")[1] if "_" in namespace else 0.0
+        relay = next((node[0] for i, node in enumerate(network.nodes(data=True), start=1) if i == int(relayid)), None)
+        if relay is None:
+            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="No such relay")
+        response = await create_track(
+            track_namespace=namespace,
+            track_dto=TrackDTO(
+                name=namespace,
+                publisher=relay,
+                delay_budget=delay_budget
+            ))
```

Comment:
Itt van mérhető gyorsulás abból, hogy a blokk aszinkron végrehajtódású is lehet? Vagy lehet, nem értem, miért kell ide ez a context manager.

Reply:
kiszedtem